src/avalanchemq/queue/durable_queue.cr

```diff
@@ -228,6 +220,24 @@ module AvalancheMQ
       @log.error { "Could not restore index: #{ex.inspect}" }
     end
 
+    # Reads all SPs in the index, truncate if there's only 0's left
+    # As we preallocate files, they are not truncated on ungraceful shutdown
+    # and thus we allocate too large memory structures in restore_index.
+    # TODO: optimize using SEEK_HOLE or compare allocated blocks * block size vs file size
+    private def truncate_sparse_file(path)
+      File.open(path, "W") do |f|
+        f.buffer_size = Config.instance.file_buffer_size
+        f.advise(File::Advice::Sequential)
+        sp = SegmentPosition.from_io file
+        if sp.zero?
+          size = file.pos - SegmentPosition::BYTESIZE
```

Comment:
When we find the end, why back up SP bytesize and truncate, doesn't that remove the last SP in the file?

Reply:
no, because we have just read a 0 SP, our pos is past that one, we don't need that one so truncate to just before the 0-SP